src/providers/watsonx.ts

```diff
@@ -0,0 +1,213 @@
+  modelName: string;
+  options: ProviderOptions;
+  env?: EnvOverrides;
+  apiKey?: string;
+  client: WatsonXAIClient | undefined;
+
+  constructor(modelName: string, options: ProviderOptions) {
+    if (!options.config) {
+      throw new Error('WatsonXProvider requires a valid config.');
+    }
+
+    const { env } = options;
+    this.modelName = modelName;
+    this.options = options;
+    this.env = env;
+    this.apiKey = this.getApiKey();
+  }
+
+  id(): string {
+    return `watsonx:${this.modelName}`;
+  }
+
+  toString(): string {
+    return `[Watsonx Provider ${this.modelName}]`;
+  }
+
+  getApiKey(): string | undefined {
+    return (
+      this.options.config.apiKey ||
+      (this.options.config.apiKeyEnvar
+        ? process.env[this.options.config.apiKeyEnvar] ||
+          this.env?.[this.options.config.apiKeyEnvar as keyof EnvOverrides]
+        : undefined) ||
+      this.env?.WATSONX_API_KEY ||
+      getEnvString('WATSONX_API_KEY')
+    );
+  }
+
+  getProjectId(): string | undefined {
+    return (
+      this.options.config.projectId ||
+      (this.options.config.projectIdEnvar
+        ? process.env[this.options.config.projectIdEnvar] ||
+          this.env?.[this.options.config.projectIdEnvar as keyof EnvOverrides]
+        : undefined) ||
+      this.env?.WATSONX_PROJECT_ID ||
+      getEnvString('WATSONX_PROJECT_ID')
+    );
+  }
+
+  getModelId(): string {
+    if (!this.modelName) {
+      throw new Error('Model name must be specified.');
+    }
+    if (this.modelName.includes(':')) {
+      const parts = this.modelName.split(':');
+      if (parts.length < 2 || !parts[1]) {
+        throw new Error(`Unable to extract modelId from modelName: ${this.modelName}`);
+      }
+      return parts[1];
+    }
+    return this.options.config.modelId || this.modelName;
+  }
+
+  async getClient(): Promise<WatsonXAIClient> {
+    if (!this.client) {
+      const apiKey = this.getApiKey();
+      if (!apiKey) {
+        throw new Error(
+          'Watsonx API key is not set. Set the WATSONX_API_KEY environment variable or add `apiKey` to the provider config.',
+        );
+      }
+
+      this.client = WatsonXAI.newInstance({
+        version: this.options.config.version || '2023-05-29',
+        serviceUrl: this.options.config.serviceUrl || 'https://us-south.ml.cloud.ibm.com',
+        authenticator: new IamAuthenticator({ apikey: apiKey }),
+      });
+    }
+    return this.client;
+  }
+
+  async callApi(prompt: string): Promise<ProviderResponse> {
+    const modelId = this.getModelId();
+    const projectId = this.getProjectId();
+    if (!modelId) {
+      throw new Error('Model ID is required for WatsonX API call.');
+    }
+    if (!this.apiKey) {
+      throw new Error(
+        'Watsonx API key is not set. Set the WATSONX_API_KEY environment variable or add `apiKey` to the provider config.',
+      );
+    }
+
+    const cache = await getCache();
+    const cacheKey = `watsonx:${this.modelName}:${prompt}`;
```

Comment:
Are you sure this will be unique enough? consider caching the whole model config